### ISSUE TITLE
Language Server: more than formatting

### ISSUE THREAD

**Issue by dpren** [avh4/elm-format#479]
This relates to discussion from https://github.com/avh4/elm-format/issues/194 about providing an AST for other tools to consume. But I'm going a step further.

Elm-format today is essentially a language server that just does formatting. I'd like to propose evolving elm-format in the direction of a more fully-featured, common language server. There are _lots_ of ways to go about this sort of thing. So I'll just present what, from my perspective, seems like the best route and we can off that. (sidenote: I'm willing to do the necessary legwork of whatever I'm talking about here)

----------------------------------------------------------------

### Why a language server?
- Performance. Complicated IDE type queries don't scale in most editor plugin environments.
- Centralizing development effort:
    - reduces work
    - improves quality accross the board
- Endures over time

----------------------------------------------------------------

### A possible vision:
1. Perhaps fork elm-format or copy parts into a new repo called, say, elm-language-server.

2. Rewrite or upgrade the existing parser's capabilities to support other desired editor features.

    I've [mentioned on discourse](https://discourse.elm-lang.org/t/error-tolerant-elm-parser-0for-editor-tooling/1066) my interest in developing error-tolerant parsing. I've started messing around with Megaparsec for this because it supports recovery and is simply a fork of Parsec, so working off elm-format's parser would be an option. Open to other suggestions.

    A question for the core team: Could elm-compiler output project-wide metadata to the disk?

    With cached info for every module, the language server would only have to worry about parsing active files.

3. Start building out features.

    How should it communicate with editors?

    One way is for the language server to just be a generic information center that editor plugins can query however they please.

    But I'm going to put forth an argument for using the Language Server Protocol.

----------------------------------------------------------------

### Why the [Language Server Protocol](https://microsoft.github.io/language-server-protocol/)?

The idea is that most common IDE features are language agnostic in terms of UI.

This provides the opportunity where:
- Languages can share one backend implementation with many editors.
- Editors can share one frontend implementation with many languages.

So this means that most of the (client) plugin work _is already done_ in all the commonly used editors.
Plugins simply extend their editor's respective implementation.
Here's an example of what that looks like in Atom:
https://github.com/atom/ide-typescript/blob/master/lib/main.js
which extends:
https://atom.io/packages/atom-ide-ui

A [Haskell LSP](https://github.com/alanz/haskell-lsp) library already exists.

__Okay. Cool short-term wins. But is this truly better for Elm in the long-term?__

__What if the LSP spec changes and becomes bad?__

Many language and editor communities are now invested in the protocol. I think any problems that arise would likely be shared problems. So far there's been a good track record of healthy cooperation.

__Will this limit the flexibility of plugin authors?__

LSP is extensible:
- Custom elm-specific features can be added at any time. Other languages have done this successfully.
- Plugins can subscribe to whatever features they want.
- Servers can choose to not implement extraneous features in the LSP spec.
- And again, LSP deals with just language agnostic features that don't demand a bespoke Elm-brand UI.

I'm interested in hearing any relevant downsides to LSP. The only other pain point I can think of is simply the fact that LSP's model is very "hands on", you're updating documents in memory and such – so it's a fair amount of backend work. But I think it results in the most consistently awesome experience.

**Comment by avh4** [comment]
Implementing LSP seems outside the scope of elm-format, but I would expect that elm-format should be a useful piece that an LSP implementation could be build on top of.  I think improving the parser is appropriate work that I would be happy to see included in elm-format.  (Also, if it makes sense to separate the parser into a separate library that both elm-format and an LSP could use, I think that would make sense too.)

To summarize, I think elm-format should remain focused on 1) formatting, and 2) converting between Elm course code and an AST format.  But I happy to accept improvements to elm-format's parser, and/or refactor elm-format to support other projects using it as a library.

**Comment by dpren** [comment]
I agree.

I wouldn't really care if the parser lives inside or outside the elm-format repo, just as long as it can be exposed separately. Reason being, an LSP would want to manage the parser state and formatting/render calls separately on its own terms.

Did you have a concrete plan for #109 yet? I think skipping erroneous top-level declarations is a good place to start. But I'd like to try and take it further to the level of sub-definitions or even expressions. This would be especially beneficial to other features like autocompletion.

Would you be open to potentially major changes to the parser from me, so long as it doesn't break anything? The recovery and incremental parsing features that I have in mind would require porting to Megaparsec. Again, I'm willing to take on the responsibility of whatever work that entails.
